### PR TITLE
Error output cjs - improvement

### PR DESCRIPTION
**Pull request by gntem**
##### Checklist

- [ ] `make -j4 test` (UNIX), or `vcbuild test` (Windows) passes
- [ ] tests and/or benchmarks are included
- [ ] documentation is changed or added
- [ ] commit message follows [commit guidelines](https://github.com/nodejs/node/blob/master/doc/guides/contributing/pull-requests.md#commit-message-guidelines)

### CODE REVIEW

File: lib/internal/errors.js [hunk 1]
```diff
@@ -1182,6 +1182,21 @@ E('ERR_UNKNOWN_ENCODING', 'Unknown encoding: %s', TypeError);
...
+
+  if(error instanceof SyntaxError && (/^import/mi).test(error.stack)){
+    const sp =  error.stack.split('\n');
+    const filepath = sp[0];
+    const line = `${sp[1]}\n${sp[2]}`
+    return `${filepath}
+${line}
+Warning: Attempted to load commonjs module but found static import statement
+Further information can be found at https://nodejs.org/api/esm.html
+SyntaxError: ${error.message}`
```
Comment (gntem) [inline lib/internal/errors.js]: the `error.stack` in this error contains the string `SyntaxError: ... ` again .. I don't know why..

File: lib/internal/modules/cjs/loader.js [hunk 2]
```diff
@@ -554,6 +555,8 @@ Module._load = function(request, parent, isMain) {
   try {
     module.load(filename);
     threw = false;
+  }catch(err){
+    throw new ESM_CJS_LOAD_ERROR(err);
   } finally {
```
Comment (gntem) [inline lib/internal/modules/cjs/loader.js:658]: is there a reason to have the `finally` here, since we catch it? this way we can remove the `threw` variable

Comment (guybedford) [reply lib/internal/modules/cjs/loader.js:658]: Yes the `finally` will still run, and is important to clear the cache for errors.

File: lib/internal/errors.js (hunk 1 above)
Comment (guybedford) [reply lib/internal/errors.js]: The text here is the error "message", where the text provided when logged to the console is a combination of `e.message` and `e.stack`.

Rather than constructing a new error as you are doing here, I would suggest moving all this code into `loader.js` as I've suggested below.

File: lib/internal/errors.js [hunk 3]
```diff
@@ -1182,6 +1182,21 @@ E('ERR_UNKNOWN_ENCODING', 'Unknown encoding: %s', TypeError);
...
 E('ERR_UNKNOWN_SIGNAL', 'Unknown signal: %s', TypeError);
+E('ESM_CJS_LOAD_ERROR', (error) => {
+
+  if(error instanceof SyntaxError && (/^import/mi).test(error.stack)){
+    const sp =  error.stack.split('\n');
+    const filepath = sp[0];
+    const line = `${sp[1]}\n${sp[2]}`
+    return `${filepath}
+${line}
+Warning: Attempted to load commonjs module but found static import statement
```
Comment (guybedford) [inline lib/internal/errors.js]: The important thing with the error message is to provide the user the **immediate** fix to get their code to run.

We are all lazy developers, we don't want to read docs, we just want to get our code running :)

So perhaps something like - 

```suggestion
Note: To load an ES module set "type": "module" in the package.json or use the .mjs extension.
```

File: lib/internal/errors.js [hunk 4]
```diff
@@ -1182,6 +1182,21 @@ E('ERR_UNKNOWN_ENCODING', 'Unknown encoding: %s', TypeError);
 E('ERR_UNKNOWN_FILE_EXTENSION', 'Unknown file extension: %s', TypeError);
 E('ERR_UNKNOWN_MODULE_FORMAT', 'Unknown module format: %s', RangeError);
 E('ERR_UNKNOWN_SIGNAL', 'Unknown signal: %s', TypeError);
+E('ESM_CJS_LOAD_ERROR', (error) => {
+
+  if(error instanceof SyntaxError && (/^import/mi).test(error.stack)){
```
Comment (guybedford) [inline lib/internal/errors.js]: So this detection would fail code like `"  import 'sdf'"` - while rare, I think we should try and be as comprehensive as possible.

It also doesn't seem to support `export default 'sdf'` or code like `export { x } from './sdf'` etc.

The **export** cases are easy to handle because we can directly check:

```js
e.message.startsWith('SyntaxError: Unexpected token export')
```

So the import regular expression check as above is really just the super tough one to think about!

I can try throw together some ideas for this too, but perhaps lets discuss this on the next review further, rather than getting bogged down in it now - let's focus on getting everything else right first I think then come back around on this.

File: lib/internal/modules/cjs/loader.js [hunk 5]
```diff
@@ -554,6 +555,8 @@ Module._load = function(request, parent, isMain) {
   try {
     module.load(filename);
     threw = false;
+  }catch(err){
+    throw new ESM_CJS_LOAD_ERROR(err);
```
Comment (guybedford) [inline lib/internal/modules/cjs/loader.js]: Here's how I'd suggest doing the error adjustment here:

```suggestion
  if (err instanceof SyntaxError) {
    checkAddEsmNote(err);
  }
  throw err;
```

and then create a separate function like the following:

```js
function checkAddEsmNote (err) {
  if (err.message.startsWith('SyntaxError: Unexpected token export') || ...importRegExCheck...) {
    // we now know it is an import / export syntax error so is likely a module
    // mutate the error to include the new info -
    err.stack = "INFO" + err.stack;
  }
}
```

**Review by guybedford — COMMENTED** [review]
This is a really excellent start! I know it feels slow and tedious and like not much progress is being made, but remember how many people this will help and make their lives easier when they hit this issue. It really is worth putting in the time so thank you for pushing through here... I think we're getting close!

**Comment by gntem** [comment]
After the latest changes.
when you load an invalid file with import statement or export
```
➜  node git:(improve-error-output) ✗ ./node --experimental-modules ./test/fixtures/es-modules/invalid.cjs
(node:27712) ExperimentalWarning: The ESM module loader is experimental.

Note: To load an ES module set "type": "module" in the package.json or use the .mjs extension.

<path>/test/fixtures/es-modules/invalid.cjs:1
import "invalid";
       ^^^^^^^^^

SyntaxError: Unexpected string
    <stack trace>
```

File: lib/internal/errors.js (hunk 4 above)
Comment (gntem) [reply lib/internal/errors.js]: what do you think about the `/^(\s|\t)*import/gmi` https://regexr.com/4i4gq ?

**Comment by guybedford** [comment]
Here's a rough suggestion re the `import` detection, these are my quick thoughts on it, if you have other ideas feel free to try them too:

1. Firstly, we need to make sure we are actually detecting the code frame and not, for example, another part of the stack that happens to use the word `import` (eg say the file name was import.js). To do this, we can do a detection for - `/^\^+$/m` which would find the `      ^^^^^^` line in the stack trace.
2. Taking the previous line, we then have the `import "invalid";` line as a source string directly, and want to apply detection to determine if this code frame corresponds to an import syntax. This could be done with a regex like `/^\s*import/` very similar like what you have actually, and would likely get us close enough to being correct (it would still fail for cases like `/*comment*/ import 'x'` or `console.log('asdf'); import 'x'` or `import    ('invalidDynamicImport)`, but those seem rare enough that this might be ok)

Those edge cases might come up in review, in which case we may have to tackle them further, but perhaps that could be ok for now.

File: test/es-module/test-esm-cjs-load-error.mjs [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+import assert from 'assert';
+import fixtures from '../common/fixtures.js';
+
+const entry = fixtures.path('/es-modules/invalid.cjs');
+
+import(entry)
+.then(assert.fail, expectsError({
+    message: `Unexpected string`,
+    stack: `
+Note: To load an ES module set "type": "module" in the package.json or use the .mjs extension.
```
Comment (guybedford) [inline test/es-module/test-esm-cjs-load-error.mjs]:
```suggestion
Note: To load an ES module using --experimental-modules set "type": "module" in the package.json or use the .mjs extension.
```
